ext/collators/revint/revint_collator.c

```diff
@@ -0,0 +1,148 @@
+ * distribute this software, either in source code form or as a compiled
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <errno.h>
+#include <stdint.h>
+#include <wiredtiger_ext.h>
+
+/*
+ * A simple WiredTiger collator for indices having a single integer key,
+ * where the ordering is descending (reversed).  This collator also
+ * requires that primary key be an integer.
+ */
+
+/* Local collator structure. */
+typedef struct {
+	WT_COLLATOR collator;		/* Must come first */
+	WT_EXTENSION_API *wt_api;	/* Extension API */
+} REVINT_COLLATOR;
+
+/*
+ * revint_compare --
+ *	WiredTiger reverse integer collation, used for tests.
+ */
+static int
+revint_compare(WT_COLLATOR *collator,
+    WT_SESSION *session, const WT_ITEM *k1, const WT_ITEM *k2, int *cmp)
+{
+	const REVINT_COLLATOR *revint_collator;
+	WT_EXTENSION_API *wtapi;
+	int ret;
+	int64_t i1, i2, p1, p2;
+
+	i1 = i2 = p1 = p2 = 0;
+	revint_collator = (const REVINT_COLLATOR *)collator;
+	wtapi = revint_collator->wt_api;
+
+	/*
+	 * All indices using this collator have an integer key, and the
+	 * primary key is also an integer. A collator is usually passed the
+	 * concatenation of index key and primary key (when available),
+	 * hence we initially unpack using "ii".
+	 *
+	 * A collator may also be called with an item that includes a index
+	 * key and no primary key.  Among items having the same index key,
+	 * an item with no primary key should sort before an item with a
+	 * primary key. The reason is that if the application calls
+	 * WT_CURSOR::search on a index key for which there are more than
+	 * one value, the search key will not yet have a primary key.  We
+	 * want to position the cursor at the 'first' matching index key so
+	 * that repeated calls to WT_CURSOR::next will see them all.
+	 *
+	 * To keep this code simple, we do not reverse the ordering
+	 * when comparing primary keys.
+	 */
+	if ((ret = wtapi->struct_unpack(wtapi, session,
```

Comment:
@agorrod, yes I considered the packing APIs, it's possible it would work, but the ext libraries are not linked against the wiredtiger library.  At least it fails to link libwiredtiger_revint_collator.so on Linux because of missing external symbols.